virtual_rainforest/core/utils.py

```diff
@@ -59,62 +61,104 @@ def check_outfile(merge_file_path: Path) -> None:
+        13, "soil", "First soil layer at -0.5 metres "
+        14, "soil", "First soil layer at -1.0 metres "
 
     Args:
-        canopy_layers: number of canopy layers soil_layers: number of soil layers
+        canopy_layers: the number of canopy layers
+        soil_layers: a list giving the depth of each soil layer as a sequence of
+            negative and strictly decreasing values.
 
     Raises:
-        InitialisationError: If the number soil or canopy layers are not both positive
-            integers
+        InitialisationError: If the number of canopy layers is not a positive
+            integer or the soil depths are not a list of strictly decreasing, negative
+            float values.
 
     Returns:
         List of canopy layer roles
     """
 
     # sanity checks for soil and canopy layers
-    if soil_layers < 1:
+    if not isinstance(soil_layers, list):
         to_raise = InitialisationError(
-            "There has to be at least one soil layer in the Virtual Rainforest!"
+            "The soil layers must be a list of layer depths."
         )
         LOGGER.error(to_raise)
         raise to_raise
 
-    if not isinstance(soil_layers, int):
-        to_raise = InitialisationError("The number of soil layers must be an integer!")
+    if len(soil_layers) < 1:
+        to_raise = InitialisationError(
+            "The number of soil layers must be greater than zero."
+        )
         LOGGER.error(to_raise)
         raise to_raise
 
-    if canopy_layers < 1:
+    if not all([isinstance(v, (float, int)) for v in soil_layers]):
+        to_raise = InitialisationError("The soil layer depths are not all numeric.")
+        LOGGER.error(to_raise)
+        raise to_raise
+
+    np_soil_layer = np.array(soil_layers)
+    if not (np.all(np_soil_layer < 0) and np.all(np.diff(np_soil_layer) < 0)):
         to_raise = InitialisationError(
-            "There has to be at least one canopy layer in the Virtual Rainforest!"
+            "Soil layer depths must be strictly decreasing and negative."
         )
         LOGGER.error(to_raise)
         raise to_raise
 
-    if canopy_layers != int(canopy_layers):
+    if not isinstance(canopy_layers, int) and not (
+        isinstance(canopy_layers, float) and canopy_layers.is_integer()
+    ):
+        to_raise = InitialisationError("The number of canopy layers is not an integer.")
+        LOGGER.error(to_raise)
+        raise to_raise
+
+    if canopy_layers < 1:
         to_raise = InitialisationError(
-            "The number of canopy layers must be an integer!"
+            "The number of canopy layer must be greater than zero."
         )
         LOGGER.error(to_raise)
         raise to_raise
 
     layer_roles = (
         ["above"]
-        + ["canopy"] * canopy_layers
+        + ["canopy"] * int(canopy_layers)
         + ["subcanopy"]
         + ["surface"]
-        + ["soil"] * soil_layers
+        + ["soil"] * len(soil_layers)
```

Comment:
I think these values should be encapsulated in an enum:

```py
class LayerRole(Enum):
    ABOVE = 1
    CANOPY = 2
    # etc.
```

Converting to and from strings isn't arduous and you could write helper methods if you like:

```py
# to string
print(LayerRole.ABOVE.name.lower())

# from string
val = getattr(LayerRole, some_string.upper())
```

Reply:
I can see the point in using the `Enum` to create a tightly controlled set of values to stop users making mistakes. Is that the main motivation for using them?

All of this should really be configured and set at a much deeper level - it's basically a core axis - so needs reworking when the data configuration is built. At that point, we'd have two core axes as DataArray dimensions: the layer number `(0, ..., 14)` and the layer roles `(1, 2, 2, ....)` or possibly `(above, canopy, canopy,...)`. 

The problem then is that you can create a DataArray with enum coordinate values but they don't serialise to file.

```python
In [35]: xarray.DataArray(
    ...:     [1,2,3,4],
    ...:     coords={
    ...:         'layer_role': [
    ...:             LayerRole['ABOVE'],
    ...:             LayerRole['ABOVE'],
    ...:             LayerRole['CANOPY'],
    ...:             LayerRole['CANOPY']
    ...:         ]
    ...:     }
    ...: ).to_netcdf("test.nc")
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)

...
ValueError: unable to infer dtype on variable 'layer_role'; xarray cannot serialize arbitrary Python objects
```
At this point, we have to start translating to and from different representations to use `Enum`. That does sound possible (https://github.com/pydata/xarray/issues/2292) but I'm going to park it for now.

Unless I've missed something more fundamental about switching now?

Reply:
@alexdewar  I'm not sure about the advantage of this, have you seen how we currently use the layer roles? Does that work with your suggestion?

Reply:
The main reason would be for safer typing. The problem with using a string is that it can be anything, so you have to remember to check in all the relevant places that the string the user has provided is a valid one. Enums are also better from a documentation standpoint.

That said, it's really not a big deal and if it requires a bunch of other changes then it's definitely not worth it -- certainly not in this PR, anyway.

Reply:
Makes a lot of sense - and I can see how they've been working over in `animals`. I'll park it for now.